reference persistence forecast implementation

wholmgren opened the pull request:
Hard to believe that persistence forecasts could be this complicated, but here we are.

The only testing I've done is to verify that `index_persistence` runs and produces the right output for a handful of inputs.

It would be worthwhile to get some feedback on the design before putting too much effort into testing. You'll find a handful of comments and questions in the diff.

Eventually closes #53.

wholmgren commented:
@alorenzo175 can you give this a high level review?

alorenzo175 commented on solarforecastarbiter/reference_forecasts/persistence.py:
```diff
@@ -0,0 +1,178 @@
+        # will raise error if you've supplied incompatible data_start,
+        # data_end, window, forecast_start, forecast_end, interval_length.
+        # Probably put in another try/except and raise better message.
+        # Or... check earlier that len(fx)==1 or len(fx)==len(fx_index)
```
yes, I think this would be more clear than this try except

alorenzo175 commented on solarforecastarbiter/reference_forecasts/main.py:
```diff
@@ -61,18 +69,138 @@ def run(site, model, init_time, start, end):
+        # time. Smarter TBD approach might be to use today's observations up
+        # until issue_time, and use yesterday's observations for issue_time
+        # until end of day. So, forecast *never* uses obs > 24 hr old at each
+        # valid time.
```
maybe too much for "reference"?

wholmgren commented:
I cleaned up the code and checked that `persistence` and `index_persistence` do the right thing. See this gist with examples: https://gist.github.com/wholmgren/3e983c8e465e17e316038ac5a9574c52

My implementation is frustratingly non-trivial because:

1. Observations can represent beginning labeled intervals, ending labeled intervals, instantaneous measurements.
2. The index persistence should account for changing solar position during interval averages.
3. The persistence quantity can be either a scalar or an array depending on what you want.

The functions in `persistence.py` are fairly general, while the run function in `main.py` makes choices based on the observations and forecast.

Still needs real tests. 

@cwhanse it would be helpful for you to review this in detail.

cwhanse commented:
The documentation helps greatly @wholmgren I understand what `window` does.  I'll complete the review

wholmgren commented:
Great. I can set up the readthedocs build for the branch in an hour or two.

wholmgren commented:
Here's a link to the rendered docs: https://wholmgren-solarforecastarbiter-core.readthedocs.io/en/benchper/reference_forecasts.html

cwhanse commented on solarforecastarbiter/reference_forecasts/persistence.py:
```diff
@@ -0,0 +1,360 @@
+
+    .. math::
+
+       m &\in \{0, 1, \ldots \frac{\textrm{data start} - \textrm{data end}}{\textrm{window}} - 1\} \\
```
Should be end - start here I think

wholmgren replied:
yes, thanks.

wholmgren commented on solarforecastarbiter/reference_forecasts/persistence.py:
```diff
@@ -0,0 +1,272 @@
+    # assumes data (possibly nan) exists from start to end
+    obs = load_data(observation, data_start, data_end)
+    # only support mean for now
+    persistence_quantity = obs.resample(window, closed=closed).mean()
```
Let me make sure I follow you. Are you saying that this behavior is wrong?

```python
In [15]: data_start, data_end = '20190412T1210', '20190412T1225'

In [16]: window = '15min'

In [17]: closed = 'left'

In [18]: s = pd.Series([1, 2, 3], index=pd.date_range(start=data_start, end=data_end, freq='5min', closed=closed))

In [19]: s
Out[19]:
2019-04-12 12:10:00    1
2019-04-12 12:15:00    2
2019-04-12 12:20:00    3
Freq: 5T, dtype: int64

In [20]: s.resample(window, closed='left').mean()
Out[20]:
2019-04-12 12:00:00    1.0
2019-04-12 12:15:00    2.5
Freq: 15T, dtype: float64
```

and that instead you'd want to see

```python
In [20]: s.resample(window, closed='left').mean()
Out[20]:
2019-04-12 12:10:00    2.0
Freq: 15T, dtype: float64
```

cwhanse replied:
If we are saying that we'll average the dta between start=12:10 and end=12:25 at 15m intervals, then I would not expect the first value to be timestamped at 12:00:00.  Using left labeling, I'd expect the first point to be timestamped 12:10 with a value of 1.33 (1 for 10 minutes, 2 for 5 minutes).

wholmgren replied:
Ok, I agree.

> We haven't discussed weighting if the window is not evenly divided by the data interval.

I'd be ok with not supporting situations in which weighting would be necessary. In the current implementation it would raise a `ValueError` (I think).

wholmgren replied:
I think this problem will be somewhat if not entirely mitigated if we pursue the additional functions as described in the [comments](https://github.com/SolarArbiter/solarforecastarbiter-core/pull/59#discussion_r275068122) below.

For the `persistence_scalar` function, we know we want to reduce the data to the mean, so we can simply call `obs.mean()`. For the `persistence_interval` function, I think we can enforce start and end times that don't cause problems. I don't think the `*_index` functions will be any different.

wholmgren commented on solarforecastarbiter/reference_forecasts/persistence.py:
```diff
@@ -0,0 +1,360 @@
+from solarforecastarbiter.reference_forecasts import forecast
+
+
+def persistence(observation, window, data_start, data_end,
```
Interesting to hear you say that because I've been kicking around similar ideas. I've also wondered if two different functions be most clear, e.g. `persistence_scalar` and `persistence_interval`.

cwhanse replied:
I like those two names better than `persistence` and `index_persistence`

wholmgren replied:
To be clear, I was thinking we'd have 4 functions total:

`persistence_scalar`
`persistence_interval`
`index_persistence_scalar`
`index_persistence_interval`

or maybe 

`persistence_scalar_index`
`persistence_interval_index`

cwhanse replied:
`persistence_scalar_index` rather than `index_persistence_scalar` seems to be easier to find.

Also, seems like a good place for a kwarg to indicate index, or no, because `persistence_scalar` and `persistence_scalar_index` will share the same signature and have many lines in common.

wholmgren replied:
Ok on the name. 

> and have many lines in common.

Not in the current implementation!

wholmgren replied:
Maybe we only need 3 functions:

1. `persistence_scalar`. Data range is used to calculate a single value to persist. Used for intraday forecasts. 
2. `persistence_interval`. Data range and interval length is used to calculate multiple values to persist. Typically used for day ahead forecasts of hourly average power. In principle could be used for intraday forecasts but less likely.
3. `persistence_scalar_index`. Data range is used to calculate a single index value to persist. Used for intraday forecasts when you want to account for solar position. 

I'm less confident about the 4th flavor:

4. `persistence_interval_index`. Data range and interval length is used to calculate multiple index values to persist. For day ahead forecasts, this is an insignificant win over `persistence_interval` because the solar position doesn't change by enough to make a significant difference in forecast accuracy. Maybe the there's a use for intraday forecasts, but I think it's less common and not sure it's worth the effort.

wholmgren commented:
Code and [example notebook](https://gist.github.com/wholmgren/3e983c8e465e17e316038ac5a9574c52) updated to reflect the discussions above. `persistence.py` now has 3 persistence functions: `persistence_scalar`, `persistence_interval`, and `persistence_scalar_index`. I've not implemented the 4th, `persistence_interval_index` because I don't think it's very useful. 

The API and the code in these more specific functions are more clear than they were in the previous general functions. And I am more confident that they are correct. The downside is that it takes a little more work to glue them together in `run_persistence`, but I think the trade off is net positive.

cwhanse commented on solarforecastarbiter/reference_forecasts/main.py:
```diff
@@ -61,18 +70,179 @@ def run(site, model, init_time, start, end):
+
+    The persistence *window* is the time over which the persistence
+    quantity (irradiance, power, clear sky index, or power index) is
+    averaged over. The persistence window is automatically determined
```
Delete 'over'

cwhanse commented on solarforecastarbiter/reference_forecasts/main.py:
```diff
@@ -61,18 +70,179 @@ def run(site, model, init_time, start, end):
+    For day ahead forecasts, only persistence of measured values
+    (*index = False*) is supported.
+
+    The persistence *window* is the time over which the persistence
```
I think it would help the docstring to first describe how `data_end` is set.  Then `window` would make more sense.

cwhanse commented on solarforecastarbiter/reference_forecasts/main.py:
```diff
@@ -61,18 +70,179 @@ def run(site, model, init_time, start, end):
+        # assumes observation.interval_length <= 1h
+        window = min(
+            pd.Timedelta('1h'),
+            max(observation.interval_length, forecast.interval_length,
```
This is a bit confusing.  What cases came up where `window = pd.Timedelta('1hr')` wouldn't work?

wholmgren replied:
`window = pd.Timedelta('1hr')` does work and is in fact what `window` is set to for the relatively common scenario that `forecast.lead_time_to_start` or `forecast.run_length` is > 1hr.

cwhanse replied:
I am suggesting that the `max` part of the window determination may be unnecessarily complicating matters.

wholmgren replied:
I'm trying to think alternatives:

1. Always use `window = 1h`. Arguably a bad approach for persistence forecasts with intervals and duration << 1h.
2. Expand the definition of a `datamodel.Forecast` so that it's predefined there. Feels like the wrong place because the `datamodel.Forecast` currently says nothing about implementation.
3. Add a `kwarg`. What's the default value? Does the script that calls `run_persistence` inspect the `observation` and `forecast` to change the default value?

I can certainly move this logic into its own function(s) that return `data_start, data_end` and that might help clarify the overall `run_persistence` function.

cwhanse replied:
Agree that the helper function would improve this code.

1. OK.  If we envision a 15min forecast with 5min intervals, agree that window = 1hr probably isn't helpful.  But it's a reference forecast, so it's more important that the forecast method is transparent and repeatable, than 'best' in some sense, in my view.

What about `min(forecast.run_length, pd.Timedelta('1hr'))`?  I'd guess that run_length would tend to be the longest of the 4 quantities in the current line.

wholmgren replied:
Agree with prioritizing transparent and repeatable. What happens if `forecast.run_length` < `observation.interval_length`? Error? Or maybe `window = max(observation.interval_length, forecast.run_length)` and forget about the `1h` limit?

cwhanse replied:
> What happens if forecast.run_length < observation.interval_length? Error? 

I would, basically that asks the forecast to downscale from the observations.  I have a bias to think of the persistence index forecast as a forecast of a mean, and so hesitate to apply the method for short times, e.g., less than 15 min, where a passing shadow will dominate the observations.

cwhanse commented on solarforecastarbiter/reference_forecasts/main.py:
```diff
@@ -61,18 +70,179 @@ def run(site, model, init_time, start, end):
+            max(observation.interval_length, forecast.interval_length,
+                forecast.lead_time_to_start, forecast.run_length)
+        )
+        data_end = issue_time
```
Implicit here is that the data continues up to `issue_time`.  Is that consistent with the meaning of `issue_time`?  I'm thinking that `issue_time` is set ahead of now to allow for processing, etc., to ensure that the 9:00am forecast is done by 9:00am.  If that's the case, then what happens if the forecast processing starts at 8:30am?

wholmgren replied:
Yes, that is implicit in the definition but I don't know what we'd do differently without further convolution of the logic. I think this is consistent with how persistence forecasts are used as a benchmark in research. I think the operational implications are beyond what we can do here. 

For calculation implications... in `persistence_scalar_index` I'm making an assumption that the `load_data` function returns `NaN` for any times that do not yet exist.

cwhanse replied:
May want to extend the window to encompass a minimum period of data.

wholmgren replied:
One option is to add a `run_time` parameter and set `data_end = run_time`. We'd still need to keep the `issue_time` parameter to determine the start/end times from the `forecast` attributes.

I don't want these functions to check the database for the most recent data and extend the window if necessary. I'd rather make a NaN forecast.

cwhanse replied:
I like the `run_time` idea.  Agree that a dynamic data window is not a good idea here.

cwhanse commented on solarforecastarbiter/reference_forecasts/main.py:
```diff
@@ -61,18 +70,179 @@ def run(site, model, init_time, start, end):
+        # until issue_time, and use yesterday's observations for issue_time
+        # until end of day. So, forecast *never* uses obs > 24 hr old at each
+        # valid time. Arguably too much for a reference forecast.
+        data_end = issue_time.floor('1d')
```
Here we're saying that a dayahead forecast is based on the midnight-to-midnight period preceding the issue time. What if the forecast start is noon, and run_length 24 hours?  The forecast would use the index at midnight yesterday for the first hour 12:00 to 1:00pm.  Seems to me that the forecast should use a 24 hour period but that hour-of-day alignment is required.

wholmgren replied:
Good catch, this would do the wrong thing in that scenario. I suggest that for now we don't support forecasts that are not midnight-to-midnight.

cwhanse commented on solarforecastarbiter/reference_forecasts/main.py:
```diff
@@ -61,18 +70,179 @@ def run(site, model, init_time, start, end):
     return resampled
+
+
+def run_persistence(observation, forecast, issue_time, index=False):
```
For a reference forecast, a primary goal needs to be transparency.  I struggled a bit to understand how this function operates.  Some additional text in the docstring would help, but also the logic for setting the data window seems convoluted.

wholmgren commented on solarforecastarbiter/reference_forecasts/persistence.py:
```diff
@@ -0,0 +1,344 @@
+
+    .. math::
+
+       GHI_{t_f} = \frac{\overline{GHI_{t_{start}} \ldots GHI_{t_{end}}}}{\overline{GHI_{{clear}_{t_{start}}} \ldots GHI_{{clear}_{t_{end}}}}} \times GHI_{{clear}_{t_f}}
```
I think ratio of the averages correctly describes the implementation for generic inputs. I believe the average of the ratios requires assertions about interval alignment that is inconsistent with the present implementation.

cwhanse replied:
Agree that the comment describes the implementation, but I disagree that it is what should be implemented.  The ratio of averages is not the average clear sky index which I thought we intended by a persistence index forecast.  Since the clear sky calculation is under our control, what is the barrier to implementation?

wholmgren replied:
No barrier, just thinking through it. I think we need to change

```python
pers_index = obs.mean() / clear_ref.mean()
```

to 

```python
clear_ref_resampled = clear_ref.resample(observation.interval_length, closed=closed, label=closed).mean()
pers_index = (obs / clear_ref_resampled).mean()
```

We probably want to ensure that the observation interval is divisible by the `clear_ref` interval. Perhaps choose `freq = pd.Timedelta('1min')` for simplicity. Otherwise we need to call interpolate, ffill, and bfill on clear_ref_resampled or accept NaNs for some fractions. Or we could find the largest factor of the observation interval length that is less than some threshold (currently 5 minutes). Neither of those is easy to explain.

cwhanse replied:
Ah, I see the wrinkle - even with average of ratios, at each interval we still compute a ratio of averages (average of data / average of clear sky) for the interval.  I think that's OK because the common understanding of irradiance is a time average over a short interval.  Same for power.  

"Otherwise …" agree these considerations are not worth the effort.

wholmgren replied:
My proposed implementation is not strictly correct for observations that are "instantaneous" (instrument sample time << interval length) and have interval length > 1 min. Say GHI is recorded for 10 seconds every 60 minutes. We want to calculate clear sky at each 60 minute point rather than the 60 minute interval average. We can handle observations that are either interval averages or instantaneous with:

```python
if observation.interval_label == 'instantaneous':
    freq = observation.interval_length
else:
    freq = pd.Timedelta('1min')
```

wholmgren commented:
Updated the example notebook to use run_time. Appears to work as expected but we'll need real tests to be sure.

alorenzo175 commented on solarforecastarbiter/reference_forecasts/forecast.py:
```diff
@@ -185,7 +185,7 @@ def interpolate(arg, freq='15min'):
         return None
     else:
-        return arg.resample(freq).interpolate()
+        return arg.resample(freq, closed=None).interpolate()
```
closed=closed

wholmgren commented:
I added a lot of tests. They are not pretty but they get the job done. 

I think the `codecov/patch` only fails because I renamed variables in the unrelated `run` function and this function isn't actually tested yet.

wholmgren commented on solarforecastarbiter/reference_forecasts/tests/test_main.py:
```diff
@@ -69,3 +72,286 @@ def test_run(model, load_forecast_return_value, site_powerplant_site_type,
+    assert data_start == pd.Timestamp('20190422T1940Z')
+    assert data_end == pd.Timestamp('20190422T1945Z')
+
+    # obs interval cannot be longer than forecast interval
```
yeah, it was bad. I've refactored ~7 tests into 45 tests.

alorenzo175 commented:
probably good to go once rebased on master

wholmgren commented:
@alorenzo175 can you check the new changes related to load_data and session? Was surprised when the requests_mock worked the first time I tried it.

alorenzo175 commented on solarforecastarbiter/reference_forecasts/main.py:
```diff
@@ -168,18 +172,23 @@ def run_persistence(observation, forecast, run_time, issue_time, index=False):
     data_start, data_end = get_data_start_end(
         observation, forecast, run_time)
 
+    def load_data():
```
doesn't load data take observation, data_start, data_end arguments? This shouldn't work with how it's called in persistence.py. And if those parameters don't matter, why bother passing the load_data function instead of just the data?

alorenzo175 replied:
the test_main functions are not thorough enough if they don't catch this me thinks

wholmgren replied:
Ah, yeah that's wrong. Test passed because I mock the calls to `persistence.py` functions. 

I'm willing to revisit the question of passing the data vs. parameters/functions to load the data, but I think we should finish this PR first.

alorenzo175 commented on solarforecastarbiter/reference_forecasts/tests/test_main.py:
```diff
@@ -327,7 +328,14 @@ def obs_5min_begin(site_metadata):
+@pytest.fixture
+def session(requests_mock):
+    session = api.APISession('TOKEN')
+    requests_mock.register_uri('GET', session.base_url)
```
to really test it, you should have the mock return the proper content for session to parse

wholmgren commented on solarforecastarbiter/reference_forecasts/tests/test_persistence.py:
```diff
@@ -0,0 +1,288 @@
+import pytest
+
+
+def load_data_base(data, observation, data_start, data_end):
```
@alorenzo175 does the actual API or `api.py` account for the interval labeling like this test function does? It looks like no. I think they should.

alorenzo175 replied:
The actual API does not and I don't think it should. You ask for values between a range and you get that range inclusive of the endpoints. No unexpected timestamps.

I guess it's ok for api.py

wholmgren replied:
Let's discuss before we do anything here.

wholmgren replied:
https://github.com/SolarArbiter/solarforecastarbiter-core/issues/80
